### PR TITLE
Support FP8 MLA prefill and 128k context.

### PR DESCRIPTION
**Pull request by weireweire**
## Motivation
MLA prefill always use 16-bit despite set `kv-cache-dtype` to fp8. We have fp8 kernel now, so this fixed it.

## Modifications
- convert qkv to fp8 to make attention faster.
- increase workspace size to make 128k runable.
- ~~remove `tile_tokens_dim` to be compatible with newer flashinfer.~~

## Accuracy Tests
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.956|±  |0.0056|
|     |       |strict-match    |     5|exact_match|↑  |0.956|±  |0.0056|
```

## Benchmarking and Profiling
my command:
```
DYN_SKIP_SGLANG_LOG_FORMATTING=1  MC_FORCE_MNNVL=1  NCCL_CUMEM_ENABLE=1  NCCL_MNNVL_ENABLE=1  PYTHONUNBUFFERED=1  SGLANG_DECODE_BOOTSTRAP_TIMEOUT=1000  SGLANG_DISABLE_TP_MEMORY_INBALANCE_CHECK=1  SGLANG_ENABLE_FLASHINFER_GEMM=true  SGLANG_ENABLE_JIT_DEEPGEMM=false  SGLANG_MOONCAKE_CUSTOM_MEM_POOL=True  SGLANG_USE_MESSAGE_QUEUE_BROADCASTER=0  TORCH_DISTRIBUTED_DEFAULT_TIMEOUT=1800  SGLANG_TORCH_PROFILER_DIR=/logs/profiles/prefill python3 -m sglang.launch_server --attention-backend trtllm_mla --chunked-prefill-size -1 --context-length 136000 --cuda-graph-max-bs 4 --data-parallel-size 1 --disable-radix-cache --disaggregation-bootstrap-port 30001 --disaggregation-transfer-backend nixl --enable-symm-mem --expert-parallel-size 1 --kv-cache-dtype fp8_e4m3 --load-balance-method round_robin --max-running-requests 4 --max-total-tokens 544000 --mem-fraction-static 0.95 --model-path /model/ --moe-dense-tp-size 1 --moe-runner-backend flashinfer_trtllm --pipeline-parallel-size 4 --quantization modelopt_fp4 --scheduler-recv-interval 1 --served-model-name deepseek-ai/DeepSeek-R1 --stream-interval 10 --tensor-parallel-size 1 --trust-remote-code --watchdog-timeout 1000000 --dist-init-addr 10.66.5.212:29500 --nnodes 1 --node-rank 0 --host 0.0.0.0
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/layers/attention/trtllm_mla_backend.py [hunk 1]
```diff
@@ -259,7 +259,7 @@ def __init__(
         self.req_to_token = model_runner.req_to_token_pool.req_to_token
 
         # Workspace allocation
-        self.workspace_size = DEFAULT_WORKSPACE_SIZE_MB * 1024 * 1024
```
Comment (ishandhanani) [inline python/sglang/srt/layers/attention/trtllm_mla_backend.py:262]: Can you maybe leave a small comment why we do `*1.2`?

File: python/sglang/srt/layers/attention/trtllm_mla_backend.py [hunk 2]
```diff
@@ -1069,6 +1069,19 @@ def forward_extend(
...
+        out = None
+        if self.data_type == torch.float8_e4m3fn:
+            out = torch.empty(
+                q.shape[0],
+                q.shape[1],
+                v.shape[2],
+                device=q.device,
+                dtype=torch.bfloat16,
+            )
+            q =q.to(torch.float8_e4m3fn)
```
Comment (nvpohanh) [inline python/sglang/srt/layers/attention/trtllm_mla_backend.py]: nit: add space after `=`

File: python/sglang/srt/layers/attention/trtllm_mla_backend.py (hunk 1 above)
Comment (weireweire) [reply python/sglang/srt/layers/attention/trtllm_mla_backend.py:262]: 128k prefill length will run out of the workspace.

File: python/sglang/srt/layers/attention/trtllm_mla_backend.py (hunk 2 above)
Comment (weireweire) [reply python/sglang/srt/layers/attention/trtllm_mla_backend.py]: fixed.

File: python/sglang/srt/layers/attention/trtllm_mla_backend.py (hunk 1 above)
Comment (hlu1) [reply python/sglang/srt/layers/attention/trtllm_mla_backend.py:262]: How did you come up with 1.2?

File: python/sglang/srt/layers/attention/trtllm_mla_backend.py [hunk 3]
```diff
@@ -1069,6 +1069,19 @@ def forward_extend(
...
+            out = torch.empty(
+                q.shape[0],
+                q.shape[1],
+                v.shape[2],
+                device=q.device,
+                dtype=torch.bfloat16,
+            )
+            q = q.to(torch.float8_e4m3fn)
+            k = k.to(torch.float8_e4m3fn)
+            v = v.to(torch.float8_e4m3fn)
```
Comment (hlu1) [inline python/sglang/srt/layers/attention/trtllm_mla_backend.py]: Is layer.scaling always 1.0? If not, do q and k need to be scaled accordingly?

File: python/sglang/srt/layers/attention/trtllm_mla_backend.py (hunk 1 above)
Comment (nvpohanh) [reply python/sglang/srt/layers/attention/trtllm_mla_backend.py:262]: Maybe we should just increase `DEFAULT_WORKSPACE_SIZE_MB` instead of the magic number 1.2?

Or should we ask FlashInfer if they can provide some util function to calculate this instead of hard-coding it in SGL code?

File: python/sglang/srt/layers/attention/trtllm_mla_backend.py [hunk 4]
```diff
@@ -1069,6 +1069,19 @@ def forward_extend(
...
             )
 
+        out = None
+        if self.data_type == torch.float8_e4m3fn:
+            out = torch.empty(
+                q.shape[0],
+                q.shape[1],
+                v.shape[2],
+                device=q.device,
+                dtype=torch.bfloat16,
```
Comment (nvpohanh) [inline python/sglang/srt/layers/attention/trtllm_mla_backend.py]: should this be hard-coded to `bfloat16` or should we just the model's dtype?

File: python/sglang/srt/layers/attention/trtllm_mla_backend.py (hunk 3 above)
Comment (nvpohanh) [reply python/sglang/srt/layers/attention/trtllm_mla_backend.py]: currently, all the known models we have uses q_scale/k_scale/v_scale = 1.0f. But if the models contain these scales, we should use them.

Could you follow the example in https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/attention/trtllm_mha_backend.py to use `layer.q_scale/k_scale/v_scale` if you need device scales or `layer.q_scale_float/k_scale_float/v_scale_float` if you need host scales?

Or just use `scaled_fp8_quant()`?
https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/quantization/fp8_kernel.py#L1429-L1463

File: python/sglang/srt/layers/attention/trtllm_mla_backend.py (hunk 1 above)
Comment (weireweire) [reply python/sglang/srt/layers/attention/trtllm_mla_backend.py:262]: Ideally we should follow https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/tensorrt_llm/common/attentionOp.cpp#L893-L899 to calculate the size. But for now I'll just increase the DEFAULT_WORKSPACE_SIZE_MB

File: python/sglang/srt/layers/attention/trtllm_mla_backend.py (hunk 3 above)
Comment (weireweire) [reply python/sglang/srt/layers/attention/trtllm_mla_backend.py]: I saw other place just use scale 1.0, like `quantize_and_rope_for_fp8`, it's called in decode and draft extend.

But I do can add a check and call `scaled_fp8_quant` with real scale or 1.0

Comment (weireweire) [reply python/sglang/srt/layers/attention/trtllm_mla_backend.py]: updated, please review

File: python/sglang/srt/layers/attention/trtllm_mla_backend.py (hunk 4 above)
Comment (weireweire) [reply python/sglang/srt/layers/attention/trtllm_mla_backend.py]: fixed

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 5]
```diff
@@ -1225,7 +1225,6 @@ def forward(self, hidden_states: torch.Tensor, topk_output: TopKOutput):
             local_expert_offset=self.moe_ep_rank * self.num_local_experts,
             local_num_experts=self.num_local_experts,
             routed_scaling_factor=self.moe_runner_config.routed_scaling_factor,
-            tile_tokens_dim=None,
```
Comment (Fridge003) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py:1229]: This argument cannot be removed until the upgrade of flashinfer

File: python/sglang/srt/layers/attention/trtllm_mla_backend.py [hunk 6]
```diff
@@ -1066,6 +1067,35 @@ def forward_extend(
                 out=torch.zeros(*output_shape, dtype=q.dtype, device=q.device),
             )
 
+        out = None
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/trtllm_mla_backend.py]: Do we need to reset the data type in the `forward_batch.attn_attend_prefix_cache=True` branch?
Since it also uses `flashinfer.prefill.trtllm_ragged_attention_deepseek` kernel

**Review by nvpohanh — APPROVED** [review]

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py (hunk 5 above)
Comment (weireweire) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py:1229]: reverted, let do that in the upgrade pr.

File: python/sglang/srt/layers/attention/trtllm_mla_backend.py (hunk 6 above)
Comment (weireweire) [reply python/sglang/srt/layers/attention/trtllm_mla_backend.py]: make sense, added fp8 support for it.

**Comment by Fridge003** [comment]
@weireweire please fix lint

**Comment by weireweire** [comment]
lint fixed

**Review by Fridge003 — APPROVED** [review]
